### PR TITLE
Simplified storage

### PR DESCRIPTION
**Pull request by LuZhang-Lou**
## Motivation

A PR contains most of https://github.com/gdanezis/off-chain-reference/pull/1 and  some other  things.
Major changes include:
1. redefine lowest level storage API, requiring the user provided DB to have `get`, `try_get`, `put`, `delete`, `isin`,  `getkeys` and`count`. `try_get` is an optimization to reduce DB access. 
2. remove `StorageList` and linkedlist stuff. Use a common key `ns` to implement the  key retrieval (i.e. `getkeys`). Note that I do not hash the base_key as done in the original PR since I like the readability of the raw key.
3. use `my_request_index` to hold pending requests from my side, then rename to `my_pending_requests`
4. add`get_my_address` API in `BusinessContext`
5. in payment_logic.py, root all StorableDict by my vasp address
6. in payment_logic.py, repurpose `reference_id_index` to store mapping to version id v.s. PaymentObject
7. refactor the object_locks code and add tests
8. consolidate `pending_commands` and `command_cache` in payment_logic.py
9. add an interface for database
10. per discussion, add tests cases to verify that processor storage is rooted by self address, and not mixed in multiple vasps. 

Note: I did not include the commit/rollback style changes in the original PR, because I think it's a big change and may need more discussions (I shared some thoughts in the email thread). Given the changes to `StorageFactory` and `StorageValue`, it turns out with the storage API changes, it is hard to keep the db entries cache , flush 
 and recovery mechanism. As I result, I deleted them from this PR, making the `atomic_writes` basically a no-op. I hope we will get a better solution (IMO locking is a good path to explore) pretty soon. But before that, we need to bear with no "atomic writes" mechanism  which I think isn't the end of world, but please shout out if you think  otherwise. 

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/website/blob/master/CONTRIBUTING.md#pull-requests)?

YES

## Test Plan

tox

(Share your test plan here. If you changed code, please provide us with clear instructions for verifying that your changes work.)

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/off-chain-reference, and link to your PR here.)

### CODE REVIEW

File: src/offchainapi/protocol.py [hunk 1]
```diff
@@ -154,47 +152,32 @@ def __init__(self, myself, other, vasp, storage, processor):
...
+            self.other_address_str, root=root
+        )
 
-        root = self.storage.make_value(self.myself.as_str(), None)
-        other_vasp = self.storage.make_value(
-            self.other_address_str, None, root=root
+        # The map of commited requests with their corresponding responses.
+        # This is also used to ensure command responses are indempotent.
+        # All requests in this store have response attached.
+        self.command_sequence = self.storage.make_dict(
```
Comment (gdanezis) [inline src/offchainapi/protocol.py:163]: The name `sequence` is really a misnomer here. What about `committed_commands` or `command_set`?

File: src/offchainapi/protocol.py [hunk 2]
```diff
@@ -337,11 +296,52 @@ def apply_response(self, request):
...
+        missing_deps = []
+        used_deps = []
+        locked_deps = []
+        for dep, lock in dep_locks.items():
+            if lock is None:
+                missing_deps.append(str(dep))
+            elif lock == LOCK_EXPIRED:
+                used_deps.append(str(dep))
+            elif lock != LOCK_AVAILABLE:
+                locked_deps.append(str(dep))
```
Comment (gdanezis) [inline src/offchainapi/protocol.py:326]: There is a final posibility, namely that the lock has the cid of a command, per the line:
self.object_locks[str(dv)] = request.cid

In that case the lock is available to the command with that cid, and not available to other commands. Is that captured here? This is to make the lock re-entrant: do we need this in this context?

File: src/offchainapi/protocol.py [hunk 3]
```diff
@@ -356,40 +356,37 @@ def sequence_command_local(self, off_chain_command):
...
+        for cv in create_versions:
+            if str(cv) in self.object_locks:
+                existing_writes.append(cv)
+        if existing_writes:
+            raise DependencyException(f'Object version already exists: {", ".join(existing_writes)}')
 
         # Ensure all storage operations are written atomically.
-        with self.rlock:
-            with self.storage.atomic_writes():
+        with self.storage.atomic_writes():
```
Comment (gdanezis) [inline src/offchainapi/protocol.py:378]: If it is a no-op you can delete it?

File: src/offchainapi/storage.py [hunk 4]
```diff
@@ -231,20 +108,15 @@ def atomic_writes(self):
         return self
 
     def __enter__(self):
-        self.rlock.acquire()
```
Comment (gdanezis) [inline src/offchainapi/storage.py:234]: Do you want to just delete the whole context manager?

File: src/offchainapi/tests/conftest.py [hunk 5]
```diff
@@ -11,6 +11,7 @@
 from ..protocol_messages import CommandRequestObject
 from ..utils import JSONFlag
 from ..crypto import ComplianceKey
+from ..sample.sample_db import SampleDB
 
 import types
 import dbm
```
Comment (gdanezis) [inline src/offchainapi/tests/conftest.py:17]: Do we still need dbm?

**Review by gdanezis — APPROVED** [review]
Nice work! Have a look at some of my comments.

General question: right now you have not ported any specific SQL backend (like the sqlite one) but rather only use the in-memory SampleDB. Do we not want to provide one persistent backend as an example, or not?

File: src/offchainapi/protocol.py (hunk 1 above)
Comment (LuZhang-Lou) [reply src/offchainapi/protocol.py:163]: makes sense! renamed to `committed_commands`

File: src/offchainapi/protocol.py (hunk 2 above)
Comment (LuZhang-Lou) [reply src/offchainapi/protocol.py:326]: good point. Looking at where this is called:
1. sequence_command_local (send a request)
2. handle_request (receive a request)

#1 and #2 are independent and the same cid shouldn't invoke both of the calls (if they really do, that looks like a bug!)  Hence IMO this shouldn't be reentrant. 

What do you think?

File: src/offchainapi/protocol.py (hunk 3 above)
Comment (LuZhang-Lou) [reply src/offchainapi/protocol.py:378]: done

File: src/offchainapi/tests/conftest.py (hunk 5 above)
Comment (LuZhang-Lou) [reply src/offchainapi/tests/conftest.py:17]: removed

File: src/offchainapi/storage.py (hunk 4 above)
Comment (LuZhang-Lou) [reply src/offchainapi/storage.py:234]: done

**Comment by LuZhang-Lou** [comment]
> Nice work! Have a look at some of my comments.
> 
> General question: right now you have not ported any specific SQL backend (like the sqlite one) but rather only use the in-memory SampleDB. Do we not want to provide one persistent backend as an example, or not?

I think it would be nice to have a sqlite impl reference, but I'm not sure whether we want to add it in this PR (it is already huge).

**Comment by gdanezis** [comment]
Awesome to see 

> > Nice work! Have a look at some of my comments.
> > General question: right now you have not ported any specific SQL backend (like the sqlite one) but rather only use the in-memory SampleDB. Do we not want to provide one persistent backend as an example, or not?
> 
> I think it would be nice to have a sqlite impl reference, but I'm not sure whether we want to add it in this PR (it is already huge).

Following a discussion with Kevin, it may be even better to not have persistent storage in the offchain component. Lets have a discussion about where persistence fits in the architecture.

Fix any conflicts and its good to land?

**Pull request merged**